kotlinx-coroutines-core/common/src/flow/DataFlow.kt

```diff
@@ -0,0 +1,335 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.flow
+
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.flow.internal.*
+import kotlinx.coroutines.internal.*
+import kotlin.coroutines.*
+import kotlin.jvm.*
+
+/**
+ * A [Flow] that contains a single updatable data [value].
+ *
+ * A data flow can be created with `DataFlow()` constructor function either without an initial value
+ * or with an initial value. Data value can be updated by setting its [value] property.
+ * Updates to the [value] are always [conflated][Flow.conflate]. So a slow collector will skip fast updates,
+ * but always collects the most recently emitted value.
+ *
+ * The name reflects the fact that [DataFlow] represents an cell in a "data flow programming" model
+ * (think of an electronic spreadsheet). Dependent values can be defined by using various operators on the
+ * flows, with [combineLatest] being especially useful to combine multiple data flows using arbitrary functions.
+ *
+ * A data flow can be [closed][close] and all its collectors will complete.
+ *
+ * A read-only interface to data flow is a [Flow]. It is supposed to be collected by the consumers of the data.
+ * The ability to read the current [value] is provided only for convenience of the code that updates the data.
```

Comment:
It is not specified how to get the current value using the `flow` _read-only_ interface, is `counter.first()` work?

May multiple `counter.first()` invocations return the same value?

Reply:
Yes. `counter.first()` does compute and return the current value.